File: backend/app/schemas/search_history.py
"""
Search History schemas for API requests and responses.
"""

from datetime import datetime
from typing import Any

from pydantic import BaseModel, Field


class SearchResultSummary(BaseModel):
    """Summary of property search result for quick display."""

    property_type: str | None = None
    bedrooms: int | None = None
    bathrooms: float | None = None
    square_footage: int | None = None
    estimated_value: float | None = None
    rent_estimate: float | None = None
    year_built: int | None = None
    thumbnail_url: str | None = None


class SearchHistoryCreate(BaseModel):
    """Schema for creating a search history entry."""

    search_query: str = Field(..., min_length=5, max_length=500)
    address_street: str | None = None
    address_city: str | None = None
    address_state: str | None = None
    address_zip: str | None = None
    property_cache_id: str | None = None
    zpid: str | None = None
    result_summary: dict[str, Any] | None = None
    search_source: str = Field(default="web", pattern="^(web|mobile|api|scanner)$")
    was_successful: bool = True
    error_message: str | None = None


class SearchHistoryResponse(BaseModel):
    """Schema for search history response."""

    id: str
    user_id: str | None
    search_query: str
    address_street: str | None
    address_city: str | None
    address_state: str | None
    address_zip: str | None
    property_cache_id: str | None
    zpid: str | None
    result_summary: dict[str, Any] | None
    search_source: str | None
    was_successful: bool
    was_saved: bool
    searched_at: datetime

    class Config:
        from_attributes = True


class SearchHistoryList(BaseModel):
    """Paginated list of search history."""

    items: list[SearchHistoryResponse]
    total: int
    limit: int
    offset: int


class SearchHistoryStats(BaseModel):
    """Statistics about user's search history."""

    total_searches: int
    successful_searches: int
    saved_from_search: int
    searches_this_week: int
    searches_this_month: int
    top_markets: list[dict[str, Any]]  # [{state: "FL", count: 15}, ...]
    recent_searches: list[SearchHistoryResponse]

File: mobile/types/searchHistory.ts
/**
 * Search History types - matching backend schemas/search_history.py exactly
 */

import { SearchSource, PaginatedResponse } from './api';

// ===========================================
// Search Result Summary
// ===========================================
export interface SearchResultSummary {
  property_type?: string | null;
  bedrooms?: number | null;
  bathrooms?: number | null;
  square_footage?: number | null;
  estimated_value?: number | null;
  rent_estimate?: number | null;
  year_built?: number | null;
  thumbnail_url?: string | null;
}

// ===========================================
// Search History Create Request
// ===========================================
export interface SearchHistoryCreate {
  search_query: string;
  address_street?: string | null;
  address_city?: string | null;
  address_state?: string | null;
  address_zip?: string | null;
  property_cache_id?: string | null;
  zpid?: string | null;
  result_summary?: SearchResultSummary | null;
  search_source?: SearchSource;
  was_successful?: boolean;
  error_message?: string | null;
}

// ===========================================
// Search History Response
// ===========================================
export interface SearchHistoryResponse {
  id: string;
  user_id: string | null;
  search_query: string;
  address_street: string | null;
  address_city: string | null;
  address_state: string | null;
  address_zip: string | null;
  property_cache_id: string | null;
  zpid: string | null;
  result_summary: SearchResultSummary | null;
  search_source: SearchSource | null;
  was_successful: boolean;
  was_saved: boolean;
  searched_at: string; // ISO datetime
}

// ===========================================
// Search History List (paginated)
// ===========================================
export type SearchHistoryList = PaginatedResponse<SearchHistoryResponse>;

// ===========================================
// Top Market Entry
// ===========================================
export interface TopMarket {
  state: string;
  count: number;
}

// ===========================================
// Search History Stats
// ===========================================
export interface SearchHistoryStats {
  total_searches: number;
  successful_searches: number;
  saved_from_search: number;
  searches_this_week: number;
  searches_this_month: number;
  top_markets: TopMarket[];
  recent_searches: SearchHistoryResponse[];
}

// ===========================================
// Search History Query Params
// ===========================================
export interface SearchHistoryQueryParams {
  limit?: number;
  offset?: number;
  successful_only?: boolean;
  source?: SearchSource;
  search?: string;
}

// ===========================================
// Helper Functions
// ===========================================

/**
 * Format search display text
 */
export function getSearchDisplayText(entry: SearchHistoryResponse): string {
  if (entry.address_street) {
    const parts = [entry.address_street];
    if (entry.address_city) parts.push(entry.address_city);
    if (entry.address_state) parts.push(entry.address_state);
    return parts.join(', ');
  }
  return entry.search_query;
}

/**
 * Get relative time string for search
 */
export function getRelativeSearchTime(searchedAt: string): string {
  const date = new Date(searchedAt);
  const now = new Date();
  const diffMs = now.getTime() - date.getTime();
  const diffMins = Math.floor(diffMs / 60000);
  const diffHours = Math.floor(diffMins / 60);
  const diffDays = Math.floor(diffHours / 24);

  if (diffMins < 1) return 'Just now';
  if (diffMins < 60) return `${diffMins}m ago`;
  if (diffHours < 24) return `${diffHours}h ago`;
  if (diffDays < 7) return `${diffDays}d ago`;
  return date.toLocaleDateString();
}
